File: dart/day5_part2_2023.dart
import 'dart:io';

class RangeMap {
  int srcStart, destStart, length;

  RangeMap(this.srcStart, this.destStart, this.length);
}

int reverseConvertNumber(int number, List<RangeMap> ranges) {
  for (int i = ranges.length - 1; i >= 0; i--) {
    RangeMap r = ranges[i];
    if (number >= r.destStart && number < r.destStart + r.length) {
      return r.srcStart + (number - r.destStart);
    }
  }
  return number;
}

bool isInSeedRanges(int number, List<List<int>> ranges) {
  for (List<int> r in ranges) {
    if (number >= r[0] && number < r[0] + r[1]) {
      return true;
    }
  }
  return false;
}

void main() {
  File file = File('input.txt');
  List<List<int>> seedRanges = [];
  List<RangeMap> currentRanges = [];
  List<List<RangeMap>> maps = [];

  file.readAsLines().then((List<String> lines) {
    for (String line in lines) {
      if (line.contains('map:')) {
        if (currentRanges.length > 0) {
          maps.add(List.from(currentRanges));
          currentRanges = [];
        }
      } else if (line.startsWith('seeds:')) {
        List<String> seedStrs = line.substring(7).split(' ');
        for (int i = 0; i < seedStrs.length; i += 2) {
          int start = int.parse(seedStrs[i]);
          int length = int.parse(seedStrs[i + 1]);
          seedRanges.add([start, length]);
        }
      } else {
        List<String> numbers = line.split(' ');
        if (numbers.length == 3) {
          int srcStart = int.parse(numbers[1]);
          int destStart = int.parse(numbers[0]);
          int length = int.parse(numbers[2]);

          currentRanges.add(RangeMap(srcStart, destStart, length));
        }
      }
    }
    if (currentRanges.length > 0) {
      maps.add(List.from(currentRanges));
    }

    int location = 0;
    while (true) {
      int seed = location;
      for (int i = maps.length - 1; i >= 0; i--) {
        seed = reverseConvertNumber(seed, maps[i]);
      }

      if (isInSeedRanges(seed, seedRanges)) {
        print(location);
        break;
      }
      location++;
    }
  });
}

File: nim/day5_part2_2023.nim
import strutils, sequtils, strformat

type RangeMap = object
  srcStart, destStart, length: int

proc reverseConvertNumber(number: int, ranges: seq[RangeMap]): int =
  for r in ranges:
    if number >= r.destStart and number < r.destStart + r.length:
      return r.srcStart + (number - r.destStart)
  return number

proc isInSeedRanges(number: int, ranges: seq[(int, int)]): bool =
  for r in ranges:
    if number >= r[0] and number < r[0] + r[1]:
      return true
  return false

when isMainModule:
  let file = open("input.txt")
  var seedRanges: seq[(int, int)]
  var currentRanges: seq[RangeMap]
  var maps: seq[seq[RangeMap]]

  for line in file.lines:
    if line.contains("map:"):
      if currentRanges.len > 0:
        maps.add(currentRanges)
        currentRanges = @[]
    elif line.startsWith("seeds:"):
      let seedStrs = line[7..^1].split(' ')
      for i in countup(0, seedStrs.len - 1, 2):
        let start = seedStrs[i].parseInt
        let length = seedStrs[i + 1].parseInt
        seedRanges.add((start, length))
    else:
      let numbers = line.split()
      if numbers.len == 3:
        let srcStart = numbers[1].parseInt
        let destStart = numbers[0].parseInt
        let length = numbers[2].parseInt
        currentRanges.add(RangeMap(srcStart: srcStart, destStart: destStart, length: length))

  if currentRanges.len > 0:
    maps.add(currentRanges)

  file.close()  # Move the file closing here

  var location = 0
  while true:
    var seed = location
    for i in countdown(maps.len - 1, 0):
      seed = reverseConvertNumber(seed, maps[i])
    if isInSeedRanges(seed, seedRanges):
      echo location
      break
    inc location